Sema: Improve availability diagnostics when `-target-min-inlining-version min` is specified

tshortli opened the pull request:
Reword diagnostics about potentially unavailable decls to improve clarity for authors of API libraries. When decls are diagnosed as potentially unavailable on an OS earlier than the deployment target, the message will now indicate that the issue would be faced by clients of the module.

For example, suppose these declarations were part of an API module named `Test` with a deployment target of 10.15 on macOS. When building `Test`, the diagnostic emitted in the body of the inlinable function `foo()` now indicates that although `bar()` is available to `Test`, it is not necessarily available to the clients of `Test`: 

```swift
@available(macOS 10.15, *)
public func bar() { ... }

@available(macOS 10.14, *)
@inlinable func foo() {
  bar() // error: 'bar' is only available in macOS 10.15 or newer; clients of 'Test' may have a lower deployment target
}
```
Resolves rdar://93466875

tshortli commented:
@swift-ci please test

xymus commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -1967,36 +1967,56 @@ void TypeChecker::checkConcurrencyAvailability(SourceRange ReferenceRange,
+    return Diagnostic(
+        IsError ? diag::availability_decl_only_version_newer_for_clients
+                : diag::availability_decl_only_version_newer_for_clients_warn,
+        D->getName(), ReferenceDC->getParentModule(), Platform, Version);
```
What is the goal of printing the reference module name here?

I'm wondering if printing it could be more confusing than helpful in the case where the unavailable decl comes from a different module. You could see a `View (as in SwiftUI.View) is only available to clients of MyLib in macOS 12.0 or newer`, when `View` is actually available more widely to the clients importing `SwiftUI` directly. Maybe we should talk about the interface of the reference module: `View is only available in the API/module interface of MyLib in macOS 12.0`.

xymus replied:
Or should we reference `D`'s module instead?

tshortli replied:
I think that using `D`'s module name would not convey what I'm trying to convey. When this diagnostic is emitted, `D` would always be available to code in the module being built. The issue is that it might not be available to the clients of the module being built. The clients of `D`'s module, however, include the module being built. So if the diagnostic were reworded to refer to `D`'s module then it is not adding any semantic meaning over the previous diagnostic wording.

tshortli replied:
I've update the diagnostic text to this:
```
'Foo' is only available in macOS X.Y or newer; clients of 'Module' may have a lower deployment target
```

xymus reviewed (approved):
I like this, it should really in understanding the new diagnostics.

tshortli commented on lib/Sema/TypeCheckAvailability.cpp:
```diff
@@ -1967,36 +1967,56 @@ void TypeChecker::checkConcurrencyAvailability(SourceRange ReferenceRange,
 
-  bool AsError = true;
   auto RequiredRange = Reason.getRequiredOSVersionRange();
-  {
```
Apparently this scope I removed had a functional purpose, fixing

tshortli commented:
@swift-ci please test

tshortli committed:
Tests: Update `attr_inlinable_available.swift` to require macOS and remove the OS versions for other platforms from availability attributes.

Upcoming changes to the diagnostics make it difficult to verify correct behavior when only matching a fraction of a diagnostic, so the test needs to run on a single platform in order to have stable diagnostics to match. The `availability_refinement_contexts_target_min_inlining.swift` test case runs on all ABI stable platforms and verifies that an appropriate version floor is used for each, so running this test suite on multiple platforms is somewhat redundant.

tshortli committed:
Tests: Add test cases for potential unavailability in class inheritance when  is specified.

tshortli committed:
Sema: Reword diagnostics about potentially unavailable decls to improve clarity for authors of API libraries. When decls are diagnosed as potentially unavailable on an OS earlier than the deployment target, the message will now indicate that the issue would be faced by clients of the module.

Resolves rdar://93466875

tshortli commented:
@swift-ci please smoke test

tshortli commented:
@swift-ci please smoke test
